File: examples/gcs-to-gke-configmaps/tests/main_test.py
# Copyright 2019 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import os
import unittest
from unittest.mock import patch
from unittest.mock import MagicMock
from unittest.mock import PropertyMock

import main


class MainTest(unittest.TestCase):
    def setUp(self):
        os.environ['OBJECT_TO_WATCH'] = "test-folder"

    @patch('main.process_finalize')
    @patch('main.create_gcs_object')
    @patch('main.create_gke_cluster')
    def test_will_process_finalize(self, mock_gke_cluster,
                                   mock_gcs_object, mock_process_finalize):
        data = {
            "bucket": "test-bucket",
            "name": "test-folder/test-file.yaml"
        }
        context = MagicMock()
        event_type = PropertyMock(
            return_value='google.storage.object.finalize')

        type(context).event_type = event_type

        main.gcs_trigger(data, context)
        assert mock_process_finalize.called

    @patch('main.process_finalize')
    @patch('main.create_gcs_object')
    @patch('main.create_gke_cluster')
    def test_will_not_process_finalize(self, mock_gke_cluster,
                                       mock_gcs_object, mock_process_finalize):
        data = {
            "bucket": "test-bucket",
            "name": "test-file.yaml"
        }
        context = MagicMock()
        event_type = PropertyMock(
            return_value='google.storage.object.finalize')

        type(context).event_type = event_type

        main.gcs_trigger(data, context)
        assert not mock_process_finalize.called

    @patch('main.process_delete')
    @patch('main.create_gcs_object')
    @patch('main.create_gke_cluster')
    def test_will_process_delete(self, mock_gke_cluster,
                                 mock_gcs_object, mock_process_delete):
        data = {
            "bucket": "test-bucket",
            "name": "test-folder/test-file.yaml"
        }
        context = MagicMock()
        event_type = PropertyMock(
            return_value='google.storage.object.delete')

        type(context).event_type = event_type

        main.gcs_trigger(data, context)
        assert mock_process_delete.called

    @patch('main.process_delete')
    @patch('main.create_gcs_object')
    @patch('main.create_gke_cluster')
    def test_will_not_process_delete(self, mock_gke_cluster,
                                     mock_gcs_object, mock_process_delete):
        data = {
            "bucket": "test-bucket",
            "name": "test-file.yaml"
        }
        context = MagicMock()
        event_type = PropertyMock(
            return_value='google.storage.object.delete')

        type(context).event_type = event_type

        main.gcs_trigger(data, context)
        assert not mock_process_delete.called

File: tools/openapi-to-cloud-armor/main_test.go
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"flag"
	"log"
	"regexp"
	"testing"

	"github.com/getkin/kin-openapi/openapi3"
	"github.com/stretchr/testify/assert"
)

func TestPathwise(t *testing.T) {
	doc, err := openapi3.NewLoader().LoadFromData([]byte(`openapi: 3.0.0
info:
  title: test
  version: '1.0'
servers:
  - url: 'http://localhost:8080'
paths:
  '/users/{name}':
    parameters:
      - schema:
          type: string
        name: name
        in: path
        required: true
    get:
      summary: Get User
      tags: []
      responses: {}
      operationId: get-user
      requestBody:
        content:
          application/json:
            schema:
              type: object
              properties: {}
          text/plain:
            schema:
              type: object
              properties: {}
        description: ''
  '/pets/{name}':
    parameters:
      - schema:
          type: string
        name: name
        in: path
        required: true
    get:
      summary: Get Pet
      tags: []
      responses: {}
      operationId: get-pet
      requestBody:
        content:
          application/json:
            schema:
              type: object
              properties: {}
          text/plain:
            schema:
              type: object
              properties: {}
        description: ''
components:
  schemas: {}`))
	if err != nil {
		log.Fatal(err)
		t.FailNow()
		return
	}
	flag.Parse()
	rules := []Rule{}
	rules, err = GeneratePathwiseRules(rules, doc)
	if err != nil {
		log.Fatal(err)
		t.FailNow()
		return
	}

	assert.Equal(t, 2, len(rules))
}

func TestMethodwise(t *testing.T) {
	doc, err := openapi3.NewLoader().LoadFromData([]byte(`openapi: 3.0.0
info:
  title: test
  version: '1.0'
servers:
  - url: 'http://localhost:8080'
paths:
  '/users/{name}':
    parameters:
      - schema:
          type: string
        name: name
        in: path
        required: true
    get:
      summary: Get User
      tags: []
      responses: {}
      operationId: get-user
      requestBody:
        content:
          application/json:
            schema:
              type: object
              properties: {}
          text/plain:
            schema:
              type: object
              properties: {}
        description: ''
  '/pets/{name}':
    parameters:
      - schema:
          type: string
        name: name
        in: path
        required: true
    get:
      summary: Get Pet
      tags: []
      responses: {}
      operationId: get-pet
      requestBody:
        content:
          application/json:
            schema:
              type: object
              properties: {}
          text/plain:
            schema:
              type: object
              properties: {}
        description: ''
components:
  schemas: {}`))
	if err != nil {
		log.Fatal(err)
		t.FailNow()
		return
	}
	flag.Parse()
	rules := []Rule{}
	rules, err = GenerateMethodwiseRules(rules, doc)
	if err != nil {
		log.Fatal(err)
		t.FailNow()
		return
	}

	assert.Equal(t, 1, len(rules))
	assert.Contains(t, rules[0].Match.Expr.Expression, "request.method=='GET'")
	assert.Contains(t, rules[0].Match.Expr.Expression, "/users/[^/]*$")
	assert.Contains(t, rules[0].Match.Expr.Expression, "/pets/[^/]*$")
}

func TestCompressed(t *testing.T) {
	doc, err := openapi3.NewLoader().LoadFromData([]byte(`openapi: 3.0.0
info:
  title: test
  version: '1.0'
servers:
  - url: 'http://localhost:8080'
paths:
  '/users/{name}':
    parameters:
      - schema:
          type: string
        name: name
        in: path
        required: true
    get:
      summary: Get User
      tags: []
      responses: {}
      operationId: get-user
      requestBody:
        content:
          application/json:
            schema:
              type: object
              properties: {}
          text/plain:
            schema:
              type: object
              properties: {}
        description: ''
  '/pets/{name}':
    parameters:
      - schema:
          type: string
        name: name
        in: path
        required: true
    get:
      summary: Get Pet
      tags: []
      responses: {}
      operationId: get-pet
      requestBody:
        content:
          application/json:
            schema:
              type: object
              properties: {}
          text/plain:
            schema:
              type: object
              properties: {}
        description: ''
  '/pets/test':
    get:
      summary: Get Pet test
      tags: []
      responses: {}
      operationId: get-pet-test
      requestBody:
        content:
          application/json:
            schema:
              type: object
              properties: {}
          text/plain:
            schema:
              type: object
              properties: {}
        description: ''
components:
  schemas: {}`))
	if err != nil {
		log.Fatal(err)
		t.FailNow()
		return
	}
	flag.Parse()
	rules := []Rule{}
	rules, err = GeneratedCompressedRules(rules, doc)
	if err != nil {
		log.Fatal(err)
		t.FailNow()
		return
	}

	assert.Equal(t, 1, len(rules))
	assert.Contains(t, rules[0].Match.Expr.Expression, "(request.method=='GET')")
	assert.Contains(t, rules[0].Match.Expr.Expression, "/users/[^/]*$")
	assert.Contains(t, rules[0].Match.Expr.Expression, "/pets/[^/]*$")
}

func TestRegex(t *testing.T) {
	regex, err := regexp.Compile(ToRegEx(`/pets/{name}`))
	if err != nil {
		t.FailNow()
	}

	assert.Equal(t, true, regex.MatchString("/pets/test123"))
	assert.Equal(t, false, regex.MatchString("/pets/test123/asd"))
	assert.Equal(t, false, regex.MatchString("/users/test123"))
	assert.Equal(t, false, regex.MatchString("/users/test123/asd"))
}

func TestFilterPaths(t *testing.T) {
	paths := []string{
		"/pet/[^/]*$",
		"/user/login$",
		"/user/[^/]*$",
		"/pet/findByStatus$",
		"/pet/findByTags$",
		"/store/inventory$",
		"/store/order/[^/]*$",
		"/user/logout$",
	}
	filtered := FilterPaths(paths)
	assert.Equal(t, 4, len(filtered))
	assert.Contains(t, filtered, "/pet/[^/]*$")
	assert.Contains(t, filtered, "/user/[^/]*$")
	assert.Contains(t, filtered, "/store/inventory$")
	assert.Contains(t, filtered, "/store/order/[^/]*$")
}
